unified-scheduler-pool/src/sleepless_testing.rs

```diff
@@ -0,0 +1,352 @@
+use std::{
+    io,
+    thread::{JoinHandle, Scope, ScopedJoinHandle},
+};
+
+#[allow(dead_code)]
+pub(crate) trait ScopeTracked<'scope>: Sized {
+    fn spawn_tracked<T: Send + 'scope>(
+        &'scope self,
+        f: impl FnOnce() -> T + Send + 'scope,
+    ) -> ScopedJoinHandle<'scope, T>;
+}
+
+pub(crate) trait BuilderTracked: Sized {
+    fn spawn_tracked<T: Send + 'static>(
+        self,
+        f: impl FnOnce() -> T + Send + 'static,
+    ) -> io::Result<JoinHandle<T>>;
+
+    #[allow(dead_code)]
+    fn spawn_scoped_tracked<'scope, 'env, T: Send + 'scope>(
+        self,
+        scope: &'scope Scope<'scope, 'env>,
+        f: impl FnOnce() -> T + Send + 'scope,
+    ) -> io::Result<ScopedJoinHandle<'scope, T>>;
+}
+
+#[cfg(not(test))]
+pub(crate) use sleepless_testing_dummy::*;
+#[cfg(test)]
+pub(crate) use sleepless_testing_real::*;
+
+#[cfg(test)]
+mod sleepless_testing_real {
+    use {
+        lazy_static::lazy_static,
+        std::{
+            cmp::Ordering::{Equal, Greater, Less},
+            collections::{HashMap, HashSet},
+            fmt::Debug,
+            sync::{Arc, Condvar, Mutex},
+            thread::{current, JoinHandle, ThreadId},
+        },
+    };
+
+    #[derive(Debug)]
+    struct Progress {
+        _name: String,
+        check_points: Vec<String>,
+        current_check_point: Mutex<String>,
+        condvar: Condvar,
+    }
+
+    #[derive(Debug)]
+    struct JustCreated;
+
+    impl Progress {
+        fn new(check_points: impl Iterator<Item = String>, name: String) -> Self {
+            let initial_check_point = format!("{JustCreated:?}");
+            let check_points = [initial_check_point.clone()]
+                .into_iter()
+                .chain(check_points)
+                .collect::<Vec<_>>();
+            let check_points_set = check_points.iter().collect::<HashSet<_>>();
+            assert_eq!(check_points.len(), check_points_set.len());
```

Comment:
seems a bit odd that we cannot have duplicate checkpoint names. If I had some sort of state-machine it would seem reasonable to me that I could use `at(NAME_OF_STATE)` to test the state-transitions.

Reply:
so true. the current impl is too primitive. this is another improvement which i'd like to have when creating a crate.

Reply:
this limitation will be lifted by this pr: https://github.com/anza-xyz/agave/pull/1797